diff --git a/src/compat-adapters/ember-source.js b/src/compat-adapters/ember-source.js
index c961d437ea2cbac64ab4c80274f3f2f3ea8d60c7..a97e7abd36c4f0959c25117d9b8cb1fa3b06e893 100644
--- a/src/compat-adapters/ember-source.js
+++ b/src/compat-adapters/ember-source.js
@@ -101,7 +101,10 @@ class default_1 extends v1_addon_1.default {
             // top-level `ember` package tries to import it until 4.10. A
             // spec-compliant ES modules implementation will treat this as a parse
             // error.
-            trees.push(new FixStringLoc([packages]));
+            trees.push(new ReplaceRequire([new FixStringLoc([packages])]));
+        }
+        else if ((0, semver_1.satisfies)(this.packageJSON.version, '<5.7.0')) {
+            trees.push(new ReplaceRequire([packages]));
         }
         if ((0, semver_1.satisfies)(this.packageJSON.version, '<5.12.0')) {
             trees.push(new FixDeprecateFunction([packages]));
@@ -159,6 +162,103 @@ class FixStringLoc extends broccoli_plugin_1.default {
         (0, fs_extra_1.outputFileSync)((0, path_1.resolve)(this.outputPath, 'ember', 'index.js'), outSource, 'utf8');
     }
 }
+class ReplaceRequire extends broccoli_plugin_1.default {
+  build() {
+      updateFileWithTransform(this, 'ember/index.js', function (babel) {
+          const { types: t } = babel;
+          function createLoader() {
+              return t.objectExpression([
+                  t.objectMethod('get', t.identifier('require'), [], t.blockStatement([
+                      t.returnStatement(t.memberExpression(t.identifier('globalThis'), t.identifier('require'))),
+                  ])),
+                  t.objectMethod('get', t.identifier('define'), [], t.blockStatement([
+                      t.returnStatement(t.memberExpression(t.identifier('globalThis'), t.identifier('define'))),
+                  ])),
+                  t.objectMethod('get', t.identifier('registry'), [], t.blockStatement([
+                      t.returnStatement(t.logicalExpression('??', t.optionalMemberExpression(t.memberExpression(t.identifier('globalThis'), t.identifier('requirejs')), t.identifier('entries'), false, true), t.optionalMemberExpression(t.memberExpression(t.identifier('globalThis'), t.identifier('require')), t.identifier('entries'), false, true))),
+                  ])),
+              ]);
+          }
+          return {
+              visitor: {
+                  CallExpression(path) {
+                      if (path.node.callee.type === 'Identifier' &&
+                          (path.node.callee.name === 'has' || path.node.callee.name === 'require') &&
+                          path.node.arguments[0].type === 'StringLiteral' &&
+                          path.node.arguments[0].value === 'ember-testing') {
+                          path.replaceWith(t.identifier('EmberTestingImpl'));
+                      }
+                  },
+                  ImportDeclaration(path) {
+                      if (path.node.source.value === 'require') {
+                          path.replaceWith(t.importDeclaration([t.importSpecifier(t.identifier('EmberTestingImpl'), t.identifier('_impl'))], t.stringLiteral('@ember/test')));
+                      }
+                  },
+                  VariableDeclaration(path) {
+                      if (path.node.declarations[0].id.type === 'Identifier' &&
+                          path.node.declarations[0].id.name === 'PartialEmber' &&
+                          path.node.declarations[0].init.type === 'ObjectExpression') {
+                          const declaration = path.node.declarations[0];
+                          const loader = declaration.init.properties.find(p => (p.type === 'ObjectProperty' && p.key.type === 'Identifier' && p.key.name) === '__loader');
+                          loader.value = createLoader();
+                      }
+                  },
+                  AssignmentExpression(path) {
+                      if (path.node.left.type === 'MemberExpression' &&
+                          path.node.left.object.type === 'Identifier' &&
+                          path.node.left.object.name === 'Ember' &&
+                          path.node.left.property.type === 'Identifier' &&
+                          path.node.left.property.name === '__loader') {
+                          path.node.right = createLoader();
+                      }
+                  },
+              },
+          };
+      });
+      replaceFile(this, '@ember/test/index.js', `export let registerAsyncHelper;
+export let registerHelper;
+export let registerWaiter;
+export let unregisterHelper;
+export let unregisterWaiter;
+export let _impl;
+
+let testingNotAvailableMessage = () => {
+throw new Error('Attempted to use test utilities, but \`ember-testing\` was not included');
+};
+
+registerAsyncHelper = testingNotAvailableMessage;
+registerHelper = testingNotAvailableMessage;
+registerWaiter = testingNotAvailableMessage;
+unregisterHelper = testingNotAvailableMessage;
+unregisterWaiter = testingNotAvailableMessage;
+
+export function registerTestImplementaiton(impl) {
+let { Test } = impl;
+registerAsyncHelper = Test.registerAsyncHelper;
+registerHelper = Test.registerHelper;
+registerWaiter = Test.registerWaiter;
+unregisterHelper = Test.unregisterHelper;
+unregisterWaiter = Test.unregisterWaiter;
+_impl = impl;
+}`);
+      replaceFile(this, 'ember-testing/index.js', `export * from './lib/public-api';
+import * as EmberTesting from './lib/public-api';
+import { registerTestImplementaiton } from '@ember/test';
+
+
+registerTestImplementaiton(EmberTesting);`);
+      replaceFile(this, 'ember-testing/lib/public-api.js', `
+export { default as Test } from './test';
+export { default as Adapter } from './adapters/adapter';
+export { default as setupForTesting } from './setup_for_testing';
+export { default as QUnitAdapter } from './adapters/qunit';
+
+import './ext/application';
+import './ext/rsvp'; // setup RSVP + run loop integration
+import './helpers'; // adds helpers to helpers object in Test
+import './initializers'; // to setup initializer`);
+  }
+}
 class FixDeprecateFunction extends broccoli_plugin_1.default {
     build() {
         let inSource = (0, fs_extra_1.readFileSync)((0, path_1.resolve)(this.inputPaths[0], '@ember', 'debug', 'index.js'), 'utf8');
@@ -223,6 +323,9 @@ function updateFileWithTransform(context, file, transformFunction) {
     }).code;
     (0, fs_extra_1.outputFileSync)((0, path_1.resolve)(context.outputPath, file), outSource, 'utf8');
 }
+function replaceFile(context, file, content) {
+  (0, fs_extra_1.outputFileSync)((0, path_1.resolve)(context.outputPath, file), content, 'utf8');
+}
 class FixCycleImports extends broccoli_plugin_1.default {
     build() {
         for (let file of ['@ember/object/observable.js', '@ember/utils/lib/is_empty.js']) {
